### PR TITLE
Add examples for binary format documentation

### LINKED ISSUE #170: Expand documentation for serialization format
Issue to improve https://github.com/typedefs/typedefs/blob/master/binaryFormat.md

Add an example typedef and example (in say Haskell) that writes/reads a term and then also show the serialisation of that term, as an example.

Related to #154 and #168, #14, #136

### PR DESCRIPTION
**Pull request by andrevidela**
fix #170

### CODE REVIEW

**Comment by andrevidela** [comment]
Ready for review, can you take a look @epost @wires ?

@fredrikNordvallForsberg can you check I did not make any mistakes in my examples?

File: docs/Binary Format/binaryFormat.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+Given this typedef:
+
+```
+(name Boolean (mu (True 1) (False 1)))
+
+(name Try (+ (var 0) (var 1)))
+
+(name LinkedList (mu (Nil 1) (Cons (* (var 1) (var 0)))))
+```
+_You can find this defintion in the `example.tdef` file_
```
Comment (fredrikNordvallForsberg) [inline docs/Binary Format/binaryFormat.md]:
```suggestion
_You can find this definition in the `example.tdef` file_
```

File: docs/Binary Format/binaryFormat.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+unknown, a description of the type could first be serialised, followed
+by the serialisation of the term (this is future work).
+
+Terms are serialised as follows:
+
+- Terms of type `T0` or `T1` do not need to be serialised --- the former does not exist, and the latter are trivial.
+- Terms of type `TSum ts` with `|ts| = 2 + k` are serialised as an integer `i` (with `i < 2 + k`, but this is not exploited), followed by the serialisation of a term of type `ts[i]`.
+- Terms of type `TProd ts` with `|ts| = 2 + k` are serialised as the serialisation of `ts[0]`, ..., `ts[1+k]`. This relies on being able to compute the width of each serialised term.
+- Terms of type `Tvar j` are serialised with a given "parameter" serialiser --- this is used to serialise `TMu`s.
+- Terms of type `Tmu ts` with |ts| = k are serialised as an integer `i` (with `i < k`, but this is not exploited), followed by the serialisation of a term of type ts[i], where the "parameter" serialiser for the new type variable is instantiated to this described procedure.
```
Comment (fredrikNordvallForsberg) [inline docs/Binary Format/binaryFormat.md]:
```suggestion
- Terms of type `TMu ts` with |ts| = k are serialised as an integer `i` (with `i < k`, but this is not exploited), followed by the serialisation of a term of type ts[i], where the "parameter" serialiser for the new type variable is instantiated to this described procedure.
```
(my original mistake, I know!)

File: docs/Binary Format/binaryFormat.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+```
+(name Boolean (mu (True 1) (False 1)))
+
+(name Try (+ (var 0) (var 1)))
+
+(name LinkedList (mu (Nil 1) (Cons (* (var 1) (var 0)))))
+```
+_You can find this defintion in the `example.tdef` file_
+
+We can generate a new haskell file that will define those type in Haskell and will create serializers and
```
Comment (fredrikNordvallForsberg) [inline docs/Binary Format/binaryFormat.md]:
```suggestion
We can generate a new Haskell file that will define those type in Haskell and will create serializers and
```

File: docs/Binary Format/binaryFormat.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+deserializers for terms of those types.
+
+We generate the haskell code using the typedef command line tool. See the [installation tutorial][INSTALL].
+
+```
+typedefs -s exmaple.tdef -d example.hs
+```
+
+You can find the generated file in this repo at `docs/Binary format/exmaple.hs`<sup>[1](#footnote)</sup>
```
Comment (fredrikNordvallForsberg) [inline docs/Binary Format/binaryFormat.md]:
```suggestion
You can find the generated file in this repo at `docs/Binary format/example.hs`<sup>[1](#footnote)</sup>
```
Also is this footnote correct?

File: docs/Binary Format/binaryFormat.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+```
+_You can find this defintion in the `example.tdef` file_
+
+We can generate a new haskell file that will define those type in Haskell and will create serializers and
+deserializers for terms of those types.
+
+We generate the haskell code using the typedef command line tool. See the [installation tutorial][INSTALL].
+
+```
+typedefs -s exmaple.tdef -d example.hs
```
Comment (fredrikNordvallForsberg) [inline docs/Binary Format/binaryFormat.md]:
```suggestion
typedefs -s example.tdef -d example.hs
```

**Comment by andrevidela** [comment]
No particular reason for the space, I'll rename it to `binary_format` the space already bit me in a very scary way haha (`rm -rf docs/Binary Format/main.o` instead of `rm -rf docs/Binary\ Format/main.o`)

File: docs/binary_format/binaryFormat.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+(name Try (+ (var 0) (var 1)))
+
+(name LinkedList (mu (Nil 1) (Cons (* (var 1) (var 0)))))
+```
+_You can find this definition in the `example.tdef` file_
+
+We can generate a new Haskell file that will define those type in Haskell and will create serializers and
+deserializers for terms of those types.
+
+We generate the haskell code using the typedef command line tool. See the [installation tutorial][INSTALL].
```
Comment (epost) [inline docs/binary_format/binaryFormat.md]:
```suggestion
We generate the haskell code using the `typedefs` command line tool. See the [installation tutorial][INSTALL].
```

File: docs/binary_format/binaryFormat.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+```
+typedefs -s example.tdef -d example.hs
+```
+
+You can find the generated file in this repo at `docs/Binary format/example.hs`<sup>[1](#footnote)</sup>
+
+## Interacting with Haskell
+
+
+In order to use typedefs, you need to import its definitions in your codebase.
```
Comment (epost) [inline docs/binary_format/binaryFormat.md]:
```suggestion
In order to use Typedefs, you need to import its definitions in your codebase.
```

File: docs/binary_format/binaryFormat.md [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+
+Here is an example of a Haskell file that imports the typedefs defintions and uses them for serializing and
```
Comment (epost) [inline docs/binary_format/binaryFormat.md]:
```suggestion
Here is an example of a Haskell file that imports the Typedefs definitions and uses them for serializing and
```

File: docs/binary_format/binaryFormat.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+ghc main.hs
+```
+
+command, and then execute the program using
+
+```
+./main
+```
+
+The output shoudl be :
```
Comment (epost) [inline docs/binary_format/binaryFormat.md]:
```suggestion
The output should be :
```

File: docs/binary_format/binaryFormat.md [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+|------------------------------|-----------|
+| `False`                      | `1`       |
+| `True`                       | `0`       |
+| `[False, True, False]`       | `111010`  |
+| `Right [False, True, False]` | `1111010` |
+
+
+##### Footnote
+
+We had to manually modify the generated haskell file in order to fix some scoping issues and add a module header.
```
Comment (epost) [inline docs/binary_format/binaryFormat.md]:
```suggestion
We had to manually modify the generated Haskell file in order to fix some scoping issues and add a module header. 
```

File: docs/binary_format/example.hs [hunk 11]
```diff
@@ -0,0 +1,96 @@
+module Example where
+
+import Data.Word
+import Data.ByteString.Lazy
+import Data.ByteString.Builder
+
+import Data.Void
+
+type Serialiser a = a -> Builder
```
Comment (epost) [inline docs/binary_format/example.hs]: Are we consistent in spelling serilaiser vs. serializer in the codebase?

File: docs/binary_format/binaryFormat.md [hunk 12]
```diff
@@ -0,0 +1,154 @@
+
+# Typedefs binary format
+
+Terms are serialised in a type-directed way, given a typedefs
```
Comment (epost) [inline docs/binary_format/binaryFormat.md]:
```suggestion
Terms are serialised in a type-directed way, given a Typedefs
```

File: docs/binary_format/binaryFormat.md [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+description --- terms make no sense on their own. If the type is
+unknown, a description of the type could first be serialised, followed
+by the serialisation of the term (this is future work).
```
Comment (epost) [inline docs/binary_format/binaryFormat.md]:
```suggestion
by the serialisation of the term. (This is future work.)
```

File: docs/binary_format/binaryFormat.md [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+```
+(name Boolean (mu (True 1) (False 1)))
+
+(name Try (+ (var 0) (var 1)))
+
+(name LinkedList (mu (Nil 1) (Cons (* (var 1) (var 0)))))
+```
+_You can find this definition in the `example.tdef` file_
+
+We can generate a new Haskell file that will define those type in Haskell and will create serializers and
```
Comment (epost) [inline docs/binary_format/binaryFormat.md]:
```suggestion
We can generate a new Haskell file that will define those types in Haskell and will create serialisers and
```

File: docs/binary_format/binaryFormat.md [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+deserializers for terms of those types.
```
Comment (epost) [inline docs/binary_format/binaryFormat.md]:
```suggestion
deserialisers for terms of those types.
```

File: docs/binary_format/binaryFormat.md [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+
+You can find the generated file in this repo at `docs/Binary format/example.hs`<sup>[1](#footnote)</sup>
+
+## Interacting with Haskell
+
+
+In order to use typedefs, you need to import its definitions in your codebase.
+
+Here is an example of a Haskell file that imports the typedefs defintions and uses them for serializing and
+deserializing terms that correspond to the specified types:
```
Comment (epost) [inline docs/binary_format/binaryFormat.md]:
```suggestion
deserialising terms that correspond to the specified types:
```

File: docs/binary_format/binaryFormat.md [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+"1"
+Just (False,"")
+"decoding succeeded"
+"1110110"
+"decoding succeeded"
+"11110110"
+"decoding succeeded"
+```
+
+Here is a table that recaps how those types are being serialized:
```
Comment (epost) [inline docs/binary_format/binaryFormat.md]:
```suggestion
Here is a table that recaps how those types are being serialised: 
```

File: docs/binary_format/example.hs (hunk 11 above)
Comment (andrevidela) [reply docs/binary_format/example.hs]: We were, until I modified the JS api. I'll make a britishified version of the API too

File: docs/binary_format/binaryFormat.md (hunk 13 above)
Comment (andrevidela) [reply docs/binary_format/binaryFormat.md]: It is not clear to me why this change is necessary. Could you explain your reasoning to me?

Comment (epost) [reply docs/binary_format/binaryFormat.md]: In this case, it's more or less that the sentence still has to make sense without parens. Alternatively, emdash (or endash, I always forget which is which) could work, e.g. "-- this is future work.", or "(which is future work)."

File: docs/binary_format/example.hs (hunk 11 above)
Comment (epost) [reply docs/binary_format/example.hs]: For some values of British. :9 🦁🇬🇧🦄

File: docs/binary_format/binaryFormat.md (hunk 13 above)
Comment (epost) [reply docs/binary_format/binaryFormat.md]: Oh, I just noticed you do use `---` in other places.

Comment (andrevidela) [reply docs/binary_format/binaryFormat.md]: I think the `---` is from fred but I'll put it here as well.

**Pull request merged**